src/referenced.cc

```diff
@@ -34,62 +35,98 @@ const int kDelimiterLength = 1;
+  if (idx >= 0) {
+    if ((unsigned int)idx >= global_words.size()) {
+      GOOGLE_LOG(ERROR) << "Global word index is too big: " << idx
+                        << " >= " << global_words.size();
+      return false;
+    }
+    *str = global_words[idx];
+  } else {
+    // per-message index is negative, its format is:
+    //    per_message_idx = -(array_idx + 1)
+    idx = -idx - 1;
+    if (idx >= reference.words_size()) {
+      GOOGLE_LOG(ERROR) << "Per message word index is too big: " << idx
+                        << " >= " << reference.words_size();
+      return false;
+    }
+    *str = reference.words(idx);
+  }
+
+  return true;
+}
+
+bool Referenced::Fill(const Attributes &attributes,
+                      const ReferencedAttributes &reference) {
+  const std::vector<std::string> &global_words = GetGlobalWords();
+  const auto &attributes_map = attributes.attributes();
+
+  for (const auto &match : reference.attribute_matches()) {
+    AttributeRef ar;
+    if (!decode(match.name(), global_words, reference, &ar.name)) {
+      return false;
+    }
+
+    const auto it = attributes_map.find(ar.name);
+    if (it != attributes_map.end()) {
+      const Attributes_AttributeValue &value = it->second;
+      if (value.value_case() == Attributes_AttributeValue::kStringMapValue) {
+        if (!decode(match.map_key(), global_words, reference, &ar.map_key)) {
+          return false;
+        }
       }
-      name = reference.words(idx);
     }
 
     if (match.condition() == ReferencedAttributes::ABSENCE) {
-      absence_keys_.push_back(name);
+      absence_keys_.push_back(ar);
     } else if (match.condition() == ReferencedAttributes::EXACT) {
-      exact_keys_.push_back(name);
+      exact_keys_.push_back(ar);
     } else if (match.condition() == ReferencedAttributes::REGEX) {
       // Don't support REGEX yet, return false to no caching the response.
       GOOGLE_LOG(ERROR) << "Received REGEX in ReferencedAttributes for "
-                        << name;
+                        << ar.name;
       return false;
     }
   }
 
+  std::sort(absence_keys_.begin(), absence_keys_.end());
+  std::sort(exact_keys_.begin(), exact_keys_.end());
+
   return true;
 }
 
-bool Referenced::Signature(const Attributes& attributes,
-                           const std::string& extra_key,
-                           std::string* signature) const {
-  const auto& attributes_map = attributes.attributes();
-  for (const std::string& key : absence_keys_) {
-    // if an "absence" key exists, return false for mis-match.
-    if (attributes_map.find(key) != attributes_map.end()) {
+bool Referenced::Signature(const Attributes &attributes,
+                           const std::string &extra_key,
+                           std::string *signature) const {
+  const auto &attributes_map = attributes.attributes();
+
+  for (const AttributeRef &key : absence_keys_) {
+    const auto it = attributes_map.find(key.name);
```

Comment:
I feel we can optimize here. Since they are sorted,  if there are multiple map keys in a string_map,
they will be next to each other,   can we only do one map lookup for multiple sub map keys?